src/types/orchestrationTypes.ts

```diff
@@ -1,13 +1,18 @@
 import { FunctionOptions, FunctionTrigger, InvocationContext } from "@azure/functions";
-import { TokenSource } from ".";
+import { Task, TokenSource } from ".";
 import { DurableOrchestrationContext } from "../classes";
 
 /**
  * Type of a Generator that can be registered as an orchestration
+ * @param T the return value of the orchestration
  */
-export type OrchestrationHandler = (
+export type OrchestrationHandler<T = any> = (
```

Comment:
I've mentioned a few times how I'm not a fan of generic types and I'd rather remove them from durable instead of add them. Please give more examples/reasoning why this `T` is helpful if you would like to add it.

Also keeping in mind that generators have these three: `T = unknown, TReturn = any, TNext = unknown`, why is TReturn the one you are passing along but not `T` or `TNext`? This change just opens up a whole design discussion that I don't think is worth it at the moment

Reply:
I think your reasoning before was to only add generics if they would prevent/detect an otherwise undetectable error. I think this addition here satisfies this criteria. It allows the user to specify the expected return value of their generator, helping to detect cases where they've returned the wrong type, or for example, neglected a path that wouldn't return anything at all. Here's an examplenpm :

```TS

// Before
const orchestrator: OrchestrationHandler = function* (context: OrchestrationContext) {
    if (Math.random() > 0.5) {
         return 'Hello';
    }
    // Suppose I expect this function to always return a string, but I neglected this path here -> error not thrown.
    return 1;
}

// After
const orchestrator: OrchestrationHandler<string> = function* (context: OrchestrationContext) {
    if (Math.random() > 0.5) {
         return 'Hello';
    }
    // Error thrown here because this path doesn't return a string
    return 1;
}
```

Basically, I consider specifying `T` the equivalent of writing `: string` at the end of a function declaration, except doing that is not valid syntax for Generator types.

To answer your other question, `T` here is determined by the SDK and is always `Task`, so that's why that change was also made. `TNext` is the result of the Durable SDKs, which unfortunately doesn't always have a consistent type, that's why I left it as `any` here. But again, it's determined by the SDK, not user code.

Reply:
I appreciate your example - that's exactly the kind of thing I'm looking for thanks. Will we be able to specify OrchestrationHandler&lt;something&gt; in our templates? What about any of our samples? I'm trying to gauge how realistically helpful this is vs theoretically helpful

Reply:
I think this change could be applied to all of our templates/samples, yes. For example, here's what the simple hello orchestration can look like now:

```TS
const orchestrator: OrchestrationHandler<string[]> = function* (context: OrchestrationContext) {
    const outputs = [];
    outputs.push(yield context.df.callActivity(activityName, 'Tokyo'));
    outputs.push(yield context.df.callActivity(activityName, 'Seattle'));
    outputs.push(yield context.df.callActivity(activityName, 'Cairo'));

    return outputs;
};
```

If we merge this, I think we should update our TS templates/samples to follow this pattern too,

Reply:
Hossam and I discussed this offline and I'll summarize here for @davidmrdavid and other's sake:

I have two main roadblocks to approval at the moment
1. It bothers me that our generics are close but not quite the same as `Generator`. The type for `Generator` is coming straight from Node.js core types and I trust Node.js designers more than I trust us. If we decide generics are worth it, I would prefer we match all three from `Generator`. Something like this:
    ```TS
    export type OrchestrationHandler<T extends Task = Task, TReturn = any, TNext = any> ...
    ```
2. As mentioned by Hossam, I usually only like generics "if they would prevent/detect an otherwise undetectable error". Initially I found Hossam's example compelling because he said you couldn't specify the return type `: string[]` on a generator, but after messing around I realized you _can_ accomplish the same thing without generics. These two are equivalent:
    ```TS
    const orchestrator: OrchestrationHandler<string[]> = function* (context: OrchestrationContext) {
    const orchestrator = function* (context: OrchestrationContext): Generator<Task, string[], any> {
    ```

Generics have caused me a lot of pain in the past and once you add them you can't take them back without it being a breaking change. That's why I'm strict about it. We can change our mind a month from now if we get user feedback asking for generics. We can't change our mind a month from now if the generics start causing problems.

Reply:
@ejizba:

For (1): our generator is only allowed to yield `Task` types, because we use `yield` as a proxy for `await`. So I think it would be a bit weird to ask users to always write `Generator<Task, ...>`in their type. Guiding users to `yield` only Tasks is the main benefit of this change, _imo_. I would be willing to compromise and allow users to parametrize `TNext` in addition to `TReturn`, but I don't think we should revert all the way to the default `Generator` type because this is a specialization of that type.

(2) I trust your experience here regarding breaking changes. Let's not rush this decision.

Given the discussion, I think there's a middle ground we could pivot to. See below.

@hossam-nasr / @ejizba : Can we consider limiting this change to simply refining `OrchestrationHandler` to be `Generator<Task, unknown, any>`? In other words, we remove the previous `unknown` with `Task` to help users guarantee that they only `yield` Tasks. Then, I believe we should be able to add generics via an overload. We can track the specific overload discussion as a post preview item. Does that seem reasonable?

Reply:
> Can we consider limiting this change to simply refining `OrchestrationHandler` to be `Generator<Task, unknown, any>`?

Yes, I am fully on board with doing this change by itself

> So I think it would be a bit weird to ask users to always write `Generator<Task, ...>`in their type.

Was your comment in regard to this suggestion?

`const orchestrator = function* (context: OrchestrationContext): Generator<Task, string[], any> {`

To me, specifying `Generator<Task, ...>` as the return type to `function*` is analogous to specifying `Promise<...>` as the return type to `async function`. Yes, the `Promise` is implied and you don't _have_ to specify a return type, but it's still best practice to do so.